File: apps/backend/test/kotlin/config/LoadSecretsTest.kt
package cloud.angora.config

import cloud.angora.constants.BackendConstants
import cloud.angora.testsupport.FakeInfisicalServer
import cloud.angora.testsupport.MapSecretsProvider
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

// The Assertions.assertThrows overload, not the reified `assertThrows<T>`: that
// one is inline and compiled for JVM 17, which the Kotlin plugin's default 1.8
// target refuses to inline. Same reason as service/AuthServiceImplTest.kt.
class LoadSecretsTest {

    private lateinit var server: FakeInfisicalServer

    @BeforeEach
    fun startServer() {
        server = FakeInfisicalServer(
            loginPath = BackendConstants.Infisical.UNIVERSAL_AUTH_LOGIN_PATH,
            secretsPath = BackendConstants.Infisical.LIST_SECRETS_PATH
        )
    }

    @AfterEach
    fun stopServer() {
        server.close()
    }

    private fun enabledEnv(vararg extra: Pair<String, String>): MapSecretsProvider =
        MapSecretsProvider(
            mapOf(
                BackendConstants.Infisical.ENABLED_ENV to BackendConstants.Infisical.ENABLED_VALUE,
                BackendConstants.Infisical.PROJECT_ID_ENV to PROJECT_ID,
                BackendConstants.Infisical.TOKEN_ENV to PRE_ISSUED_TOKEN,
                BackendConstants.Infisical.DOMAIN_ENV to server.domain
            ) + extra
        )

    @Test
    fun `returns the environment provider untouched when the flag is off`() {
        val env = MapSecretsProvider(mapOf("TOKEN" to "from-env"))

        val secrets = loadSecrets(env)

        assertSame(env, secrets)
        assertTrue(server.requests.isEmpty())
    }

    @Test
    fun `treats any value other than true as disabled`() {
        val env = MapSecretsProvider(
            mapOf(
                BackendConstants.Infisical.ENABLED_ENV to "1",
                "TOKEN" to "from-env"
            )
        )

        assertSame(env, loadSecrets(env))
    }

    @Test
    fun `prefers an Infisical value over the same name in the environment`() {
        server.serveSecrets("TOKEN" to "from-infisical")

        val secrets = loadSecrets(enabledEnv("TOKEN" to "from-env"))

        assertEquals("from-infisical", secrets.get("TOKEN"))
    }

    @Test
    fun `falls back to the environment for a name Infisical does not define`() {
        server.serveSecrets("TOKEN" to "from-infisical")

        val secrets = loadSecrets(enabledEnv("OTHER" to "only-in-env"))

        assertEquals("only-in-env", secrets.get("OTHER"))
        assertNull(secrets.get("ABSENT"))
        assertEquals("fallback", secrets.get("ABSENT", "fallback"))
    }

    @Test
    fun `skips the login call when a pre-issued token is configured`() {
        server.serveSecrets("TOKEN" to "from-infisical")

        loadSecrets(enabledEnv())

        assertNull(server.requestFor(BackendConstants.Infisical.UNIVERSAL_AUTH_LOGIN_PATH))
        assertEquals(
            "${BackendConstants.Infisical.BEARER_PREFIX}$PRE_ISSUED_TOKEN",
            server.requestFor(BackendConstants.Infisical.LIST_SECRETS_PATH)?.authorization
        )
    }

    @Test
    fun `logs in with client credentials and uses the returned access token`() {
        server.serveSecrets("TOKEN" to "from-infisical")
        val env = MapSecretsProvider(
            mapOf(
                BackendConstants.Infisical.ENABLED_ENV to BackendConstants.Infisical.ENABLED_VALUE,
                BackendConstants.Infisical.PROJECT_ID_ENV to PROJECT_ID,
                BackendConstants.Infisical.DOMAIN_ENV to server.domain,
                BackendConstants.Infisical.CLIENT_ID_ENV to "client-id",
                BackendConstants.Infisical.CLIENT_SECRET_ENV to "client-secret"
            )
        )

        val secrets = loadSecrets(env)

        assertEquals("from-infisical", secrets.get("TOKEN"))

        val login = server.requestFor(BackendConstants.Infisical.UNIVERSAL_AUTH_LOGIN_PATH)
        assertEquals("POST", login?.method)
        assertTrue(login?.body?.contains("client-secret") == true)

        assertEquals(
            "${BackendConstants.Infisical.BEARER_PREFIX}${FakeInfisicalServer.ISSUED_TOKEN}",
            server.requestFor(BackendConstants.Infisical.LIST_SECRETS_PATH)?.authorization
        )
    }

    @Test
    fun `targets the configured project, environment and path`() {
        val secretPath = "/backend"
        loadSecrets(
            enabledEnv(
                BackendConstants.Infisical.ENVIRONMENT_ENV to "prod",
                BackendConstants.Infisical.SECRET_PATH_ENV to secretPath
            )
        )

        val query = server.requestFor(BackendConstants.Infisical.LIST_SECRETS_PATH)?.query
        assertTrue(query?.contains("${BackendConstants.Infisical.PROJECT_ID_QUERY}=$PROJECT_ID") == true)
        assertTrue(query?.contains("${BackendConstants.Infisical.ENVIRONMENT_QUERY}=prod") == true)
        assertTrue(query?.contains(BackendConstants.Infisical.SECRET_PATH_QUERY) == true)
    }

    @Test
    fun `aborts startup when the project id is missing`() {
        val env = MapSecretsProvider(
            mapOf(BackendConstants.Infisical.ENABLED_ENV to BackendConstants.Infisical.ENABLED_VALUE)
        )

        val error = assertThrows(IllegalStateException::class.java) { loadSecrets(env) }

        assertTrue(error.message?.contains(BackendConstants.Infisical.PROJECT_ID_ENV) == true)
    }

    @Test
    fun `aborts startup when neither a token nor client credentials are configured`() {
        val env = MapSecretsProvider(
            mapOf(
                BackendConstants.Infisical.ENABLED_ENV to BackendConstants.Infisical.ENABLED_VALUE,
                BackendConstants.Infisical.PROJECT_ID_ENV to PROJECT_ID
            )
        )

        val error = assertThrows(IllegalStateException::class.java) { loadSecrets(env) }

        assertTrue(error.message?.contains(BackendConstants.Infisical.CLIENT_ID_ENV) == true)
    }

    @Test
    fun `aborts startup rather than serving environment values when the fetch fails`() {
        server.secretsStatus = 403
        server.secretsBody = """{"message":"forbidden"}"""

        val error = assertThrows(IllegalStateException::class.java) {
            loadSecrets(enabledEnv("TOKEN" to "from-env"))
        }

        assertTrue(error.message?.contains("403") == true)
    }

    @Test
    fun `names the domain when Infisical cannot be reached at all`() {
        // Port 1 on loopback: nothing listens, so the connection is refused.
        val unreachable = "http://127.0.0.1:1"
        val env = MapSecretsProvider(
            mapOf(
                BackendConstants.Infisical.ENABLED_ENV to BackendConstants.Infisical.ENABLED_VALUE,
                BackendConstants.Infisical.PROJECT_ID_ENV to PROJECT_ID,
                BackendConstants.Infisical.TOKEN_ENV to PRE_ISSUED_TOKEN,
                BackendConstants.Infisical.DOMAIN_ENV to unreachable
            )
        )

        val error = assertThrows(IllegalStateException::class.java) {
            loadSecrets(env)
        }

        assertEquals(
            BackendConstants.Infisical.Failures.UNREACHABLE.format(unreachable),
            error.message
        )
    }

    @Test
    fun `aborts startup when the login response carries no access token`() {
        val env = MapSecretsProvider(
            mapOf(
                BackendConstants.Infisical.ENABLED_ENV to BackendConstants.Infisical.ENABLED_VALUE,
                BackendConstants.Infisical.PROJECT_ID_ENV to PROJECT_ID,
                BackendConstants.Infisical.DOMAIN_ENV to server.domain,
                BackendConstants.Infisical.CLIENT_ID_ENV to "client-id",
                BackendConstants.Infisical.CLIENT_SECRET_ENV to "client-secret"
            )
        )
        server.loginBody = """{"unexpected":true}"""

        val error = assertThrows(IllegalStateException::class.java) { loadSecrets(env) }

        assertEquals(
            BackendConstants.Infisical.Failures.MALFORMED_LOGIN_RESPONSE,
            error.message
        )
    }

    private companion object {
        const val PROJECT_ID = "proj-123"
        const val PRE_ISSUED_TOKEN = "pre-issued-token"
    }
}

File: packages/secrets/src/loadSecrets.test.ts
import { describe, expect, it, vi } from 'vitest'
import { loadSecrets } from './loadSecrets.js'
import { INFISICAL_CONFIG, INFISICAL_ROUTES } from './constants.js'
import type { EnvSource } from './infisical.js'

const PROJECT_ID = 'proj-123'

function enabledEnv(extra: EnvSource = {}): EnvSource {
  return {
    [INFISICAL_CONFIG.ENABLED_ENV]: INFISICAL_CONFIG.ENABLED_VALUE,
    [INFISICAL_CONFIG.PROJECT_ID_ENV]: PROJECT_ID,
    [INFISICAL_CONFIG.TOKEN_ENV]: 'pre-issued-token',
    ...extra,
  }
}

function jsonResponse(body: unknown, status = 200): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'Content-Type': 'application/json' },
  })
}

function secretsFetch(secrets: Record<string, string>) {
  return vi.fn(async () =>
    jsonResponse({
      secrets: Object.entries(secrets).map(([secretKey, secretValue]) => ({
        secretKey,
        secretValue,
      })),
    }),
  )
}

describe('loadSecrets with Infisical disabled', () => {
  it('reads straight from the environment and never calls out', async () => {
    const fetchImpl = vi.fn()
    const secrets = await loadSecrets({ TOKEN: 'from-env' }, fetchImpl)

    expect(secrets.get('TOKEN')).toBe('from-env')
    expect(fetchImpl).not.toHaveBeenCalled()
  })

  it('returns the fallback for a name the environment does not define', async () => {
    const secrets = await loadSecrets({})

    expect(secrets.get('MISSING')).toBeUndefined()
    expect(secrets.get('MISSING', 'default')).toBe('default')
  })

  it('treats any value other than "true" as disabled', async () => {
    const fetchImpl = vi.fn()
    const secrets = await loadSecrets(
      { [INFISICAL_CONFIG.ENABLED_ENV]: '1', TOKEN: 'from-env' },
      fetchImpl,
    )

    expect(secrets.get('TOKEN')).toBe('from-env')
    expect(fetchImpl).not.toHaveBeenCalled()
  })
})

describe('loadSecrets with Infisical enabled', () => {
  it('prefers an Infisical value over the same name in the environment', async () => {
    const fetchImpl = secretsFetch({ TOKEN: 'from-infisical' })
    const secrets = await loadSecrets(
      enabledEnv({ TOKEN: 'from-env' }),
      fetchImpl,
    )

    expect(secrets.get('TOKEN')).toBe('from-infisical')
  })

  it('falls back to the environment for a name Infisical does not define', async () => {
    const fetchImpl = secretsFetch({ TOKEN: 'from-infisical' })
    const secrets = await loadSecrets(
      enabledEnv({ OTHER: 'only-in-env' }),
      fetchImpl,
    )

    expect(secrets.get('OTHER')).toBe('only-in-env')
  })

  it('skips the login call when a pre-issued token is configured', async () => {
    const fetchImpl = secretsFetch({ TOKEN: 'from-infisical' })
    await loadSecrets(enabledEnv(), fetchImpl)

    expect(fetchImpl).toHaveBeenCalledTimes(1)
    expect(String(fetchImpl.mock.calls[0]?.[0])).toContain(
      INFISICAL_ROUTES.LIST_SECRETS,
    )
  })

  it('logs in with client credentials and uses the returned access token', async () => {
    const fetchImpl = vi
      .fn()
      .mockResolvedValueOnce(jsonResponse({ accessToken: 'issued-token' }))
      .mockResolvedValueOnce(
        jsonResponse({
          secrets: [{ secretKey: 'TOKEN', secretValue: 'from-infisical' }],
        }),
      )

    const env = enabledEnv({
      [INFISICAL_CONFIG.TOKEN_ENV]: undefined,
      [INFISICAL_CONFIG.CLIENT_ID_ENV]: 'client-id',
      [INFISICAL_CONFIG.CLIENT_SECRET_ENV]: 'client-secret',
    })
    const secrets = await loadSecrets(env, fetchImpl)

    expect(secrets.get('TOKEN')).toBe('from-infisical')
    expect(String(fetchImpl.mock.calls[0]?.[0])).toContain(
      INFISICAL_ROUTES.UNIVERSAL_AUTH_LOGIN,
    )
    const listHeaders = (fetchImpl.mock.calls[1]?.[1] as RequestInit).headers
    expect(listHeaders).toMatchObject({ Authorization: 'Bearer issued-token' })
  })

  it('targets the configured project, environment and path', async () => {
    const fetchImpl = secretsFetch({})
    await loadSecrets(
      enabledEnv({
        [INFISICAL_CONFIG.ENVIRONMENT_ENV]: 'prod',
        [INFISICAL_CONFIG.SECRET_PATH_ENV]: '/backend',
      }),
      fetchImpl,
    )

    const url = new URL(String(fetchImpl.mock.calls[0]?.[0]))
    expect(url.origin).toBe(INFISICAL_CONFIG.DEFAULT_DOMAIN)
    expect(url.searchParams.get('projectId')).toBe(PROJECT_ID)
    expect(url.searchParams.get('environment')).toBe('prod')
    expect(url.searchParams.get('secretPath')).toBe('/backend')
  })

  it('honors a self-hosted domain, trailing slash and all', async () => {
    const fetchImpl = secretsFetch({})
    await loadSecrets(
      enabledEnv({
        [INFISICAL_CONFIG.DOMAIN_ENV]: 'https://infisical.internal/',
      }),
      fetchImpl,
    )

    expect(String(fetchImpl.mock.calls[0]?.[0])).toContain(
      `https://infisical.internal${INFISICAL_ROUTES.LIST_SECRETS}`,
    )
  })
})

describe('loadSecrets failure handling', () => {
  it('rejects instead of falling back when the project id is missing', async () => {
    await expect(
      loadSecrets(
        { [INFISICAL_CONFIG.ENABLED_ENV]: INFISICAL_CONFIG.ENABLED_VALUE },
        vi.fn(),
      ),
    ).rejects.toThrow(INFISICAL_CONFIG.PROJECT_ID_ENV)
  })

  it('rejects when neither a token nor client credentials are configured', async () => {
    await expect(
      loadSecrets(
        enabledEnv({ [INFISICAL_CONFIG.TOKEN_ENV]: undefined }),
        vi.fn(),
      ),
    ).rejects.toThrow(INFISICAL_CONFIG.CLIENT_ID_ENV)
  })

  it('rejects when the secrets request fails, without serving env values', async () => {
    const fetchImpl = vi.fn(async () => jsonResponse({ message: 'nope' }, 403))

    await expect(
      loadSecrets(enabledEnv({ TOKEN: 'from-env' }), fetchImpl),
    ).rejects.toThrow('403')
  })

  it('rejects when the login request fails', async () => {
    const fetchImpl = vi.fn(async () => jsonResponse({}, 401))

    await expect(
      loadSecrets(
        enabledEnv({
          [INFISICAL_CONFIG.TOKEN_ENV]: undefined,
          [INFISICAL_CONFIG.CLIENT_ID_ENV]: 'client-id',
          [INFISICAL_CONFIG.CLIENT_SECRET_ENV]: 'client-secret',
        }),
        fetchImpl,
      ),
    ).rejects.toThrow('401')
  })

  it('rejects a response with no secrets array', async () => {
    const fetchImpl = vi.fn(async () => jsonResponse({ unexpected: true }))

    await expect(loadSecrets(enabledEnv(), fetchImpl)).rejects.toThrow(
      /secrets array/,
    )
  })
})
